lib/Firtool/Firtool.cpp

```diff
@@ -0,0 +1,246 @@
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInjectDUTHierarchyPass());
+
+  pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+      firrtl::createLowerCHIRRTLPass());
+
+  // Run LowerMatches before InferWidths, as the latter does not support the
+  // match statement, but it does support what they lower to.
+  pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+      firrtl::createLowerMatchesPass());
+
+  // Width inference creates canonicalization opportunities.
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferWidthsPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(
+      firrtl::createMemToRegOfVecPass(opt.replSeqMem, opt.ignoreReadEnableMem));
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferResetsPass());
+
+  if (opt.exportChiselInterface) {
+    if (opt.chiselInterfaceOutDirectory.empty()) {
+      pm.nest<firrtl::CircuitOp>().addPass(createExportChiselInterfacePass());
+    } else {
+      pm.nest<firrtl::CircuitOp>().addPass(createExportSplitChiselInterfacePass(
+          opt.chiselInterfaceOutDirectory));
+    }
+  }
+
+  if (opt.dedup)
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createDedupPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createWireDFTPass());
+
+  if (opt.vbToBV) {
+    pm.addNestedPass<firrtl::CircuitOp>(firrtl::createLowerFIRRTLTypesPass(
+        firrtl::PreserveAggregate::All, firrtl::PreserveAggregate::All));
+    pm.addNestedPass<firrtl::CircuitOp>(firrtl::createVBToBVPass());
+  }
+
+  if (!opt.lowerMemories)
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        firrtl::createFlattenMemoryPass());
+
+  // The input mlir file could be firrtl dialect so we might need to clean
+  // things up.
+  pm.addNestedPass<firrtl::CircuitOp>(firrtl::createLowerFIRRTLTypesPass(
+      opt.preserveAggregate, firrtl::PreserveAggregate::None));
+  // Only enable expand whens if lower types is also enabled.
+  auto &modulePM = pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>();
+  modulePM.addPass(firrtl::createExpandWhensPass());
+  modulePM.addPass(firrtl::createSFCCompatPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInlinerPass());
+
+  // Preset the random initialization parameters for each module. The current
+  // implementation assumes it can run at a time where every register is
+  // currently in the final module it will be emitted in, all registers have
+  // been created, and no registers have yet been removed.
+  if (opt.isRandomEnabled(FirtoolOptions::RandomKind::Reg))
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        firrtl::createRandomizeRegisterInitPass());
+
+  if (opt.useOldCheckCombCycles) {
+    if (opt.preserveAggregate == firrtl::PreserveAggregate::None)
+      pm.nest<firrtl::CircuitOp>().addPass(firrtl::createCheckCombCyclesPass());
+    else
+      emitWarning(module.getLoc())
+          << "CheckCombCyclesPass doens't support aggregate "
```

Comment:
To be resolved separately, but can't help but notice:
```suggestion
          << "CheckCombCyclesPass doesn't support aggregate "
```